Tasks/RunJMeterLoadTest/Start-ApacheJMeterTest.ps1

```diff
@@ -0,0 +1,412 @@
+                {
+                    if ($type.typeName -eq "ThresholdMessage")
+                    {
+                        Write-Warning ( "[{0}] {1} occurrences of {2} " -f $type.typeName, $errorDetail.occurrences, $errorDetail.messageText)
+                    }
+                    else
+                    {
+                        Write-Warning ( "[{0}] {1} occurrences of ['{2}','{3}','{4}'] : {5} " -f $type.typeName, $errorDetail.occurrences, $errorDetail.scenarioName, $errorDetail.testCaseName,
+                        $subType.subTypeName, $errorDetail.messageText)
+                    }
+                }
+                if ($type.typeName -eq "ThresholdMessage" -and $subType.subTypeName -eq "Critical" -and $global:MonitorThresholds -and $subType.occurrences -gt $ThresholdLimit)
+                {
+                    Write-Error ( "Your loadtest has crossed the permissible {0} threshold violations with {1} violations" -f $ThresholdLimit, $subType.occurrences )
+                }
+            }
+        }
+    }
+}
+
+function CheckTestErrors($headers, $run)
+{
+    if ($global:MonitorThresholds)
+    {
+        $uri = [String]::Format("{0}/_apis/clt/testruns/{1}/errors?type=ThresholdMessage&detailed=True&{2}", $global:ElsAccountUrl, $run.id, $apiVersion)
+        $errors = Invoke-RestMethod -ContentType "application/json" -UserAgent $userAgent -Uri $uri -Headers $headers
+
+        if ($errors -and $errors.count -gt 0 -and  $errors.types.count -gt 0)
+        {
+            foreach ($subType in $errors.types.subTypes)
+            {
+                if ($subType.subTypeName -eq 'Critical' -and $subType.occurrences -gt $ThresholdLimit)
+                {
+                    $global:ThresholdExceeded = $true
+                    return $true;
+                }
+            }
+        }
+    }
+    return $false;
+}
+
+function ShowMessages($headers, $run)
+{
+    $uri = [String]::Format("{0}/_apis/clt/testruns/{1}/messages?{2}", $global:ElsAccountUrl, $run.id, $apiVersion)
+    $messages = Invoke-RestMethod -ContentType "application/json" -UserAgent $userAgent -Uri $uri -Headers $headers
+    if ($messages)
+    {
+        $sMessages = $messages.value | Sort-Object loggedDate
+        foreach ($message in $sMessages)
+        {
+            switch ($message.messageType)
+            {
+                "info"     { Write-Host -NoNewline ("[Message]{0}" -f $message.message) }
+                "output"   { Write-Host -NoNewline ("[Message]{0}" -f $message.message) }
+                "warning"  { Write-Warning $message.message }
+                "error"    { Write-Error $message.message }
+                "critical" { Write-Error $message.message }
+            }
+        }
+    }
+}
+
+function UploadTestDrop($testdrop, $src)
+{
+    $dest = $testdrop.accessData.dropContainerUrl
+    $sas = $testdrop.accessData.sasKey
+
+    $azcopy = Get-ToolPath -Name "AzCopy\AzCopy.exe"
```

Comment:
There is ongoing effort to reduce dependency of Task with Agent. You may want to consider bundling all your tools within your task there by reducing any dependency.